site/en/docs/extensions/mv3/geolocation/index.md

```diff
@@ -0,0 +1,206 @@
+}
+
+// getCurrentPosition returns a prototype based object, so the properties
+// end up being stripped off when sent over to our service worker. To get
+// around this, we deeply clone it
+function clone(obj) {
+  const copy = {};
+  // Return the value of any non true object (typeof(null) is "object") directly.
+  // null will throw an error if you try to for/in it. We can just return
+  // the value early.
+  if (obj === null || !(obj instanceof Object)) {
+    return obj;
+  } else {
+    for (const p in obj) {
+      copy[p] = clone(obj[p]);
+    }
+  }
+  return copy;
+}
+
+async function getLocation() {
+  // we use a raw Promise here so we can pass `resolve` and `reject` into the
+  // callbacks for getCurrentPosition
+  return new Promise((resolve, reject) => {
+    navigator.geolocation.getCurrentPosition(
+      (loc) => resolve(clone(loc)),
+      // in case the user doesnt have/is blocking `geolocation`
+      (err) => reject(err)
+    );
+  });
+}
+```
+
+
+With that in place, we are now ready to access the Offscreen Document in our service worker!
+
+```js
+chrome.offscreen.createDocument({
+  url: 'offscreen.html',
+  reasons: [chrome.offscreen.Reason.GEOLOCATION || chrome.offscreen.Reason.DOM_SCRAPING],
+  justification: 'geolocation access',
+});
+```
+
+Note that when you access an offscreen document, you need to include a `reason`. The `geolocation` reason was not originally available, so we specify a fallback of `DOM\_SCRAPING`, and explain in our `justification` section what the code is actually doing. This information is used by the Chrome Web Store's review process to ensure offscreen documents are being used for a valid purpose.
+
+Once you have a reference to the Offscreen Document, you can send it a message to ask for it to give you updated geolocation information.
+
+{% Label %}service\_worker.js:{% endLabel %}
+```js
+const OFFSCREEN_DOCUMENT_PATH = '/offscreen.html';
+
+chrome.runtime.onMessage.addListener(handleMessages);
+
+let creating; // A global promise to avoid concurrency issues
+
+async function hasDocument() {
+  // Check all windows controlled by the service worker to see if one
+  // of them is the offscreen document with the given path
+  const offscreenUrl = chrome.runtime.getURL(OFFSCREEN_DOCUMENT_PATH);
+  const matchedClients = await clients.matchAll();
+
+  return matchedClients.some(c => c.url === offscreenUrl)
+}
+
+async function setupOffscreenDocument(path) {
+  //if we do not have a document, we are already setup and can skip
+  if (!(await hasDocument())) {
+    // create offscreen document
+    if (creating) {
+      await creating;
+    } else {
+      creating = chrome.offscreen.createDocument({
+        url: path,
+        reasons: [chrome.offscreen.Reason.GEOLOCATION || chrome.offscreen.Reason.DOM_SCRAPING],
+        justification: 'add justification for geolocation use here',
+      });
+
+      await creating;
+      creating = null;
+    }
+  }
+}
+
+async function getGeolocation() {
```

Comment:
Can you please also address this: " Might also be worth quickly mentioning why we need hasDocument and setupOffscreendoc (this can be short with a link to the offscreen docs)."

Reply:
sorry, meant to ask - there are brief comments there currently, and I do not believe I can link to content from within a comment - are you wanting me to add an inline reference to `https://developer.chrome.com/docs/extensions/reference/offscreen/#example` or add an additional section before/after the sample that explains those steps?

Reply:
One approach could be to add an aside before the code snippet saying something like:

"A Chrome extension can only have one open offscreen document at a time and offscreen documents should be closed once they are no longer needed. To learn more about how to create and close offscreen documents, checkout the offscreen documentation."